# **Support Vector Machines for Cancer Detection**

# **1. Introduction**

This report explores the application of Decision Tree (DT) and Support Vector Machine (SVM) classifiers to the breast cancer detection problem using the breast_cancer dataset from sklearn.datasets. The objective is to develop an SVM model, optimize its hyperparameters, and compare its performance against a Decision Tree model. Evaluation metrics will be used to assess predictive accuracy, and a discussion will be provided to determine which model is better suited for this classification task. Additionally, DTs and SVMs will be further explored in detail, including their key parameters, techniques, and approaches for making predictions on new data.

# **2. Decision Trees**


**2.1 How Decision Trees for Regression differ from Classification**

Decision Trees for regression and classification differ significantly when it comes to output, splitting criteria, and predictions.

**Classification Trees:**

* **Output:** Target variable is categorical (e.g., “Malignant” or “Benign”).
* **Splitting:** Splits are chosen to maximize information gain (e.g., Gini impurity, entropy).
* **Prediction:** Leaf nodes predict class labels or class probabilities.

**Regression Trees:**

* **Output:** Target variable is continuous (e.g., tumor size, house price).

* **Splitting:** Splits are chosen to minimize variance (mean squared error) in the target variable.

* **Prediction:** Leaf nodes predict a numeric value, usually the mean of the target values in that leaf.

To put it simply, the key difference is that classification predicts categories and regression predicts numbers.



**2.2 How Decision Trees Predict New Data Points**

1. Start at the root node.
2. Compare the new data point’s feature with the node’s split condition. The condition can be numerical feature (e.g., "Is Age > 30?") or a categorial feature (e.g., "Is Outlook = Sunny?").
3. Follow the left or right child depending on the comparison.
4. Repeat until a leaf node is reached.
5. Return the prediction from the leaf:
* **Classification:** majority class in that leaf.
* **Regression:** average target value of samples in that leaf.


In [84]:
# # Example_______________________________
#             [Age <= 30?]
#              /      \
#            Yes       No
#          [Class A]   [Income <= 50k?]
#                      /       \
#                  Class B     Class C


Suppose we have a person with Age = 35 and Income = 40k.

Step 1: Root node Age <= 30? → 35 > 30 → go right.

Step 2: Next node Income <= 50k? → 40k <= 50k → go left.

Step 3: Reach leaf node → predict Class B.

# **3. Support Vector Machines (SVMs)**


**2.1 Optimization Criterion of SVMs**

During training, Support Vector Machines (SVMs) optimize the criterion of maximizing the margin, which is the distance between the separating hyperplane and the closest data points from each class (called support vectors). Maximizing this margin helps the model create the most robust decision boundary for classifying new data.

Maximizing the Margin Means:

* **Optimal Hyperplane:** SVMs find the hyperplane that best separates the classes.
* **Distance to Data:** The chosen hyperplane is the farthest from the nearest data points of any class.
* **Support Vectors:** These closest points determine the position and orientation of the hyperplane and thus define the margin.

By maximizing the margin, SVMs improve generalization and reduce the chance of misclassifying new data points.


**2.2 How SVMs Handle Non-Linearly Separable Data (Kernel Trick)**

A linear SVM can only separate data with a straight line (or flat hyperplane). When the data is not linearly separable, we can imagine “lifting” it into a higher-dimensional space, where it may become possible to separate the classes with a straight boundary.

**Problem with high dimensions:**

* Explicitly transforming all the data into a higher-dimensional space can be very computationally expensive.
* For example, a simple 2D dataset might become 10-dimensional after the transformation.
* Calculating and storing all these extra features would be inefficient.

**The Kernel Trick:**

The SVM kernel trick enables Support Vector Machines to perform non-linear classification by implicitly projecting data into a higher-dimensional space without explicitly computing the transformation. Rather than transforming the data and calculating dot products in that space, a kernel function computes these dot products directly in the original, lower-dimensional space, making the process more efficient.

This allows SVMs to:

* Handle very high-dimensional spaces efficiently.
* Learn non-linear decision boundaries without explicitly computing all the extra features.
* Use different kernels (linear, polynomial, RBF) depending on the problem.

The kernel trick lets SVMs separate non-linear data efficiently without the heavy computation of explicitly mapping it to a higher-dimensional space.

**2.3 How SVMs Handle Multi-Class Classification**

SVMs are inherently binary classifiers, meaning they can only separate two classes at a time. To handle problems with more than two classes, SVMs can be extended using strategies that combine multiple binary classifiers:

1. **One-vs-Rest (OvR):**

* Train one SVM for each class, treating that class as positive and all others as negative.
* For a new data point, predict the class whose classifier outputs the highest confidence.

2. **One-vs-One (OvO):**

* Train one SVM for each pair of classes.
* For a new data point, each classifier “votes” for a class.
* The class with the most votes is chosen as the prediction.

Multi-class SVMs break the problem into multiple binary classification tasks and combine their results to predict among multiple classes.

# **4. Breast Cancer Detection/Prediction**

# **4.1 Dataset Description**



The breast cancer dataset is a classic dataset for binary classification provided by `sklearn.datasets`.

- **Number of samples:** 569  
- **Number of features:** 30 numeric features  
- **Target classes:**  
  - Malignant (0) – cancerous  
  - Benign (1) – non-cancerous  

- **Feature details:** The features are computed from digitized images of fine needle aspirates (FNA) of breast masses and describe characteristics of the cell nuclei, such as mean radius, texture, perimeter, area, smoothness, compactness, concavity, symmetry, and fractal dimension.

- **Dataset properties:**  
  - Already cleaned and preprocessed, with no missing values.  
  - Widely used to train and evaluate machine learning models for cancer detection.

- **Source:** [Wisconsin Diagnostic Breast Cancer (WDBC) dataset](https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29)


In [85]:
from sklearn.datasets import load_breast_cancer
breast_cancer_dataset = load_breast_cancer()
breast_cancer_dataset.keys()


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [86]:
print("terget names:", breast_cancer_dataset.target_names)
print("data shape:", breast_cancer_dataset.data.shape)
print("feature names:", breast_cancer_dataset.feature_names)
print("number of samples:", breast_cancer_dataset.target.shape[0])
breast_cancer_dataset.target


terget names: ['malignant' 'benign']
data shape: (569, 30)
feature names: ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
number of samples: 569


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

# **4.2 SVM Model Parameters and Hyperparameters**

In **sklearn.svm.SVC**, the main parameters and hyperparameters that affect model performance are:

- **kernel**: Defines the type of decision boundary.  
  Options: linear, poly, rbf (default), sigmoid.  
  **Tuning**: To tune this parameter use linear for linearly separable data; and rbf or poly for non-linear data.

- **C (Regularization / Soft Margin)**: Controls the trade-off between maximizing the margin and minimizing misclassification.  
  Large C → stricter margin, fewer misclassifications, but has the risk of overfitting.  
  Small C → softer margin, allows misclassifications, and allows for better generalization.  
  **Tuning**: To tune this parameter, test multiple values using grid search (e.g., C = [0.1, 1, 10, 100]).

- **gamma** (for rbf, poly, sigmoid kernels): Defines the influence of a single training sample.  
  Small gamma → smoother decision boundary.  
  Large gamma → more complex boundary, risk of overfitting.  
  **Tuning**: Test multiple values (e.g., gamma = [0.001, 0.01, 0.1, 1, scale]).

- **degree** (for poly kernel): Degree of the polynomial function (default 3).

- **Other parameters**:  
  class_weight → handle imbalanced datasets.  
  probability=True → enable probability estimates (slower).

**Tuning for Optimal Performance**:  
- Use **grid search or random search with cross-validation** to find the best combination of C, gamma, and kernel.  
- Start with simple settings (kernel=linear, C=1.0, gamma=scale) and gradually explore more complex options.  
- **Soft margin** (controlled by C) allows SVM to handle real-world data that is not perfectly linearly separable, balancing margin width and misclassification tolerance.


# **4.3 Relevant Kernel Functions, Their Role, and Justification**

Support Vector Machines (SVMs) use **kernel functions** to transform the input data into higher-dimensional spaces, allowing the model to find a suitable decision boundary even when data is not linearly separable.

**Common Kernel Functions**

- **Linear Kernel**  
  - Does not transform the data; finds a straight hyperplane to separate classes.  
  - Fast to compute and easy to interpret.  
  - Best for data that is approximately linearly separable, like the breast cancer dataset.  

- **Polynomial Kernel**  
  - Maps data into a higher-dimensional polynomial feature space.  
  - Can model interactions between features and capture curved decision boundaries.  
  - Requires tuning the `degree` parameter to control complexity.  

- **Radial Basis Function (RBF) Kernel**  
  - Maps data into an infinite-dimensional space.  
  - Flexible and capable of modeling complex, non-linear patterns.  
  - Requires careful tuning of the `gamma` parameter, which controls the influence of each training sample.  

- **Sigmoid Kernel**    
  - Less commonly used for tabular datasets, but can be used for certain non-linear problems.

**Role of Kernels**

- Kernels allow SVMs to handle **non-linear relationships** without explicitly computing transformations into high-dimensional spaces.  
- The **choice of kernel determines the shape of the decision boundary**:  
  - Linear → straight hyperplane  
  - Polynomial → curved boundaries of polynomial degree  
  - RBF → flexible, locally adapted boundaries  

**Justification for Kernel Selection in This Implementation**

- The kernel selected is a linear kernel, because the **breast cancer dataset** has 30 numeric features that are roughly linearly separable.  
- Using a **linear kernel**:  
  - Provides high accuracy and low computational cost.  
  - Avoids unnecessary complexity, which reduces the risk of overfitting.  
  - Produces a decision boundary that is interpretable and stable.  

- More complex kernels, like RBF or polynomial, do not improve performance significantly for this dataset and may require additional hyperparameter tuning.


For breast cancer classification, the linear kernel is the most appropriate choice, balancing performance, simplicity, and interpretability.


# **4.4 SVM Model Parameters Explanation**

**C (Regularization Parameter):** The C parameter controls the trade-off between a smooth decision boundary and correctly classifying training points. A smaller C value results in a wider margin and a simpler model, which can prevent overfitting. A larger C value creates a narrower margin and a more complex model that tries to classify all training points correctly, which can lead to overfitting.

**decision_function_shape='ovr' (One-vs-Rest):** This parameter is used for multi-class classification. The "ovr" (One-vs-Rest) strategy trains a separate binary classifier for each class against all other classes. For example, in a three-class problem (A, B, C), it would train a classifier for A vs. (B and C), another for B vs. (A and C), and a third for C vs. (A and B). The final prediction is based on the highest score from these classifiers.

**degree (Polynomial Kernel Degree):** The degree parameter is specific to the polynomial kernel. It sets the degree of the polynomial function used to create a non-linear decision boundary. A higher degree allows for a more complex, curved boundary, which can fit the training data more closely but increases the risk of overfitting.

**gamma (Kernel Coefficient):** The gamma parameter is a coefficient for the rbf, poly, and sigmoid kernels. It defines the influence of a single training example. A low gamma value means a large influence, creating a smoother boundary. A high gamma value means a small influence, causing the model to focus on only nearby points and potentially creating a very complex, overfitted boundary.

**kernel (Kernel Type):** The kernel parameter specifies the type of function used to transform the data. This "kernel trick" allows SVM to find non-linear decision boundaries without explicitly mapping the data to a higher dimension. Common options include:

* `linear`: For linearly separable data.

* `poly`: For a polynomial boundary.

* `rbf`: A versatile choice for non-linear data.

* `sigmoid`: A function often used in neural networks.

# **Spliting the data train/test**

In [87]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_dataset.data,breast_cancer_dataset.target, test_size = 0.2, random_state = 42)
print(X_train.shape)
print(X_test.shape)


(455, 30)
(114, 30)


# **4.5 Implementing SVM Classifier**
For this implementaition a linear kernel is used.

In [88]:
# import support vector classification model
from sklearn.svm import SVC
# Create a SVM model
model = SVC(kernel='linear')
# Train the SVM model
model.fit(X_train, y_train)

SVC(kernel='linear')

# **SVM Classifier Results Evaluation**
The model has a 95.6% accuracy.

In [89]:
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
# Make predictions
y_pred = model.predict(X_test)
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print("Precision:",metrics.precision_score(y_test, y_pred, average ='weighted'))
print("Recall:",metrics.recall_score(y_test, y_pred, average = 'weighted'))
print("F1-score:",metrics.f1_score(y_test, y_pred, average = 'weighted'))

print(classification_report(y_test, y_pred, target_names=breast_cancer_dataset.target_names))


Accuracy: 0.956140350877193
Precision: 0.9569049312470365
Recall: 0.956140350877193
F1-score: 0.9558014271241044
              precision    recall  f1-score   support

   malignant       0.97      0.91      0.94        43
      benign       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



# **Decision Tree Implementation**

In [90]:
 #--- Train with Decision Tree ---#
from sklearn.tree import DecisionTreeClassifier
# Build the model
model = DecisionTreeClassifier()
# Train the model
model.fit(X_train, y_train)
# Make predictions
y_pred = model.predict(X_test)


# **Result**
The model has a 93.8% accuracy.

In [91]:

print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print("Precision:",metrics.precision_score(y_test, y_pred, average ='weighted'))
print("Recall:",metrics.recall_score(y_test, y_pred, average = 'weighted'))
print("F1-score:",metrics.f1_score(y_test, y_pred, average = 'weighted'))
report = classification_report(y_test, y_pred, target_names =breast_cancer_dataset.target_names)
print(report)


Accuracy: 0.9385964912280702
Precision: 0.9390179995443153
Recall: 0.9385964912280702
F1-score: 0.938731642017737
              precision    recall  f1-score   support

   malignant       0.91      0.93      0.92        43
      benign       0.96      0.94      0.95        71

    accuracy                           0.94       114
   macro avg       0.93      0.94      0.93       114
weighted avg       0.94      0.94      0.94       114



# **4.6 Comparison of SVM-BC and DT Model**

- **Overall Accuracy:**  
  - SVM: **95.6%**  
  - Decision Tree: **93.9%**  
  - Observation: SVM correctly predicts more cases overall.

- **Class-wise Performance:**  
  - **Malignant Class:**  
    - Precision: SVM **0.97** vs DT **0.91**  
    - Recall: SVM **0.91** vs DT **0.93**  
    - Interpretation: SVM has fewer false positives; DT detects slightly more actual malignant cases.  
  - **Benign Class:**  
    - Precision: SVM **0.95** vs DT **0.96**  
    - Recall: SVM **0.99** vs DT **0.94**  
    - Interpretation: SVM correctly identifies almost all benign cases; DT has slightly fewer false positives.

- **F1-Scores and Averages:**  
  - Macro F1: SVM **0.9558** vs DT **0.9387**  
  - Weighted F1: SVM **0.956** vs DT **0.94**  
  - Observation: SVM balances precision and recall more effectively across both classes.


**Conclusion:**  
The SVM-BC model is better suited for this breast cancer dataset because it achieves higher overall accuracy and F1-scores compared to the Decision Tree. It balances precision and recall effectively across both malignant and benign classes, ensuring fewer misclassifications, which is crucial in a medical diagnosis scenario. Therefore, SVM-BC provides more reliable and accurate predictions for this problem.


